packages/interfaces/src/metadata.ts

```diff
@@ -1,11 +1,25 @@
 import type { PeerId } from "@libp2p/interface";
 
 import type { ShardInfo } from "./enr.js";
-import type { IBaseProtocolCore, ShardingParams } from "./protocols.js";
+import type {
+  IBaseProtocolCore,
+  ProtocolError,
+  ShardingParams
+} from "./protocols.js";
+
+export type QueryResult =
```

Comment:
nit: it should be generic and should cover `PreparePushMessageResult` as well

proposed structure:
```ts
// Most common generic to cover potential cases in future
type Result<T, E> =
  | {
      query: T;
      error: null;
    }
  | {
      query: null;
      error: E;
    };

export type ProtocolResult = Result<ShardInfo, ProtocolError>; // or QueryResult 
```

Reply:
The generic should already be able to cover that. Note: this PR is only for `metadata` and LightPush will be handled in a separate PR. Thanks for the flag though!